lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp

```diff
@@ -400,6 +400,32 @@ class StableHLOToTTIRCompareOpConversionPattern
     // rewriter may miss replacing all uses due to different output tensor type.
     // Replacing all uses of srcOp explicitly.
     rewriter.replaceAllOpUsesWith(srcOp, TTIROp);
+    return success();
+  }
+};
+
+class StableHLOToTTIRSelectOpConversionPattern
+    : public OpConversionPattern<mlir::stablehlo::SelectOp> {
+  using OpConversionPattern<mlir::stablehlo::SelectOp>::OpConversionPattern;
+
+public:
+  LogicalResult
+  matchAndRewrite(mlir::stablehlo::SelectOp srcOp,
+                  mlir::stablehlo::SelectOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    auto outputType = mlir::cast<RankedTensorType>(
+        getTypeConverter()->convertType(srcOp.getResult().getType()));
+    tensor::EmptyOp outputTensor = rewriter.create<tensor::EmptyOp>(
+        srcOp.getLoc(), outputType.getShape(), outputType.getElementType());
+
+    rewriter.replaceOpWithNewOp<::mlir::tt::ttir::WhereOp>(
```

Comment:
I think so, it's still a 1:1 mapping, just with a different name. What I don't think we should be doing in the translation layer is decomposing into multiple ops (i.e. `select -> where if dim==0 else select -> some_other_op`) or decomposing into multiple ops (i.e. `select -> transpose + where` but curious to hear others' opinion.

Reply:
> I think so, it's still a 1:1 mapping, just with a different name. What I don't think we should be doing in the translation layer is decomposing into multiple ops (i.e. `select -> where if dim==0 else select -> some_other_op`) or decomposing into multiple ops (i.e. `select -> transpose + where` but curious to hear others' opinion.

This sounds good, but I'm not sure if we can manage to do so for every case in indexing op types.. This will be mostly driven by design from TTNN. If they manage to do 1:1 mapping with PyTorch, we'll probably be able to do the same. However, I'm not certain if this is going to happen sooner or later.. 

In any case, we should force clean mapping from the start. And depending on TTNN achievements, we should probably discuss further strategies for all frontends :))

Reply:
Sounds good, thanks guys!

Reply:
We have come across a couple of cases where a 1-1 mapping is not possible, so for that we decided to create custom passes that runs after the StableHLOToTTIR conversion but before the TTIRToTTNN conversion. 

- First case is pattern matching for converting 3 ops (constantOp, broadcastOp and maximumOp) to a single ReluOp. 

- Second case Defne is working on where the stablehlo.gatherOp will first be converted to ttir.gatherOp and then converted to ttir.embeddingsOP in a custom pass. 

As this list grows and we have multiple of these custom passes, maybe should convert them all into a single pass, so it becomes a two step conversion:

```
FrontEnd Dialect -> TTIR 1-1 Conversion
TTIR 1-1 Conversion -> TTIR Native Dialect (Legalization for all Thin Conversions) 
TTIR Native Dialect -> TTIRToTTNN Dialect
```

Reply:
@uazizTT I think most of these cases will be covered with decomposition and fusing passes on the TTIR level, just to reflect all backends, not only stable hlo. But I think this is still TBD/WIP :))

Reply:
> * First case is pattern matching for converting 3 ops (`constantOp`, `broadcastOp` and `maximumOp`) to a single `ReluOp`.

I think this pattern match should have probably been done in TTIR.  We can port it.

> * Second case Defne is working on where the `stablehlo.gatherOp` will first be converted to `ttir.gatherOp` and then converted to `ttir.embeddingsOP` in a custom pass.

This one sounds OK

Reply:
@jserbedzijaTT is adding a decompose pass in TTIR:
https://github.com/tenstorrent/tt-mlir/pull/969

We should utilize that single place for decompositions.

Similarly, we need to come up with the fusing pass as well:
https://github.com/tenstorrent/tt-mlir/issues/1059

@uazizTT can you please help me understand why we need this specific fusion right now? I am trying to understand the priority for its need right now, so we can come up with the proper solution for it.

Reply:
I've just now caught up with this whole conversation, let me try to summarize the key points here because I would like to share this with the whole third party crew so we are all in sync. **Please everyone let me know if you would add or change something.**
- TTIR dialect should be general enough to cover 1:1 mappings from all third party dialects.
- In case when the same op is differently named between dialects, TTIR dialect op name should be chosen as the most common name used in popular third party dialects, to facilitate adoption. For example if all third party dialects use `select`, we shouldn't use `where` only because torch and metal ttnn use `where`. TTNN dialect names on the other hand should match the metal ttnn ops names.
- Situations where multiple simpler ops can be fused into one complex op should be handled like this:
  - We should add the complex op to TTIR dialect too, so it is general enough to support direct lowering from another dialect which has such op.
  - We should not implement fusion directly during conversion from specific third party dialect to TTTIR, because then we would have to write such fusion for each third party dialect. Instead, we should implement it inside fusion pass in TTIR (#1059).
- In situations where third party dialect supports some complex op which is not supported by metal ttnn (and they don't plan to support it), we need to decompose it into set of supported ttnn ops (if possible). In that case we still want to create the said complex op in TTIR dialect and to convert from third party dialect into TTIR 1:1, for all reasons about generality stated above, but then the decomposition should be implemented inside decomposition pass in TTIR (#969).
  - One exception here is when op is supported by metal ttnn but it doesn't currently work or is lacking some functionality, and we can decompose it into set of different ops which are working fine. That should not be implemented inside decomposition pass! Instead, that should be implemented inside different pass used for workarounds (this pass is currently WIP). Before doing any workarounds, the procedure for Metal issues should be followed first. (I'll follow up with that separately and link it here).
- This should cover most cases, any other situation that arises should be reviewed on case by case basis so we can refine these guidelines.

Unresolving thread for more visibility.

Reply:
> In case when the same op is differently named between dialects, TTIR dialect op name should be chosen as the most common name used in popular third party dialects, to facilitate adoption. For example if all third party dialects use select, we shouldn't use where only because torch and metal ttnn use where. TTNN dialect names on the other hand should match the metal ttnn ops names.

I partially disagree with this. From my perspective, TTIR should be as wide as possible, and users from any framework should use a set of ops that can be directly mapped from refferent framework. From MLIR end, we should map that kind of ops to the common set we further use (e.g. wide vs narrow TTIR).

Reply:
How is the part you've quoted in clash with that? You think we should have both select and where op in TTIR? That seems too excessive to me tbh, TTIR would just become brainless mashup of all third party dialects jumbled together. No other dialect has duplicated ops, duplicated ops = duplicated code and it introduces confusion to the user which op should be used.

Reply:
Personal preference, but it makes more sense to me to have both Select and Where ops. 

This way, folks coming from different frameworks will have an easier time adjusting to TTIR ops. In the background, we can do aliases and actually fallback to use only a single op. E.g. from TTIR wide (Select, Where, etc.) to TTIR narrow (Select). 

I don't see how and why this introduces confusion and brainless mashup for all third parties. As mentioned, it enables our users to to have easier time fitting in into TTIR.

Reply:
> This way, folks coming from different frameworks will have an easier time adjusting to TTIR ops.

Not necessarily, you are trying to save them time finding out which op in our dialect to use, but on the other hand you will lose them time trying to figure out why do we have both select and where op and which one should they use. Imagine if every third party dialect used different name, you would have ten functionally same ops in TTIR with different names. So from the user side I don't think you are doing them any good, and then on our side you are making our code more messy. Every dialect has specs and since we are providing 1:1 mapping for each op it should be fairly easy for them to figure out which op to use.

Reply:
@mrakitaTT you nicely summarized everything. It would be ideal if we could capture your summary somewhere, though I’m not certain if our public docs are the right spot for it.

On the topic of having one or multiple TTIR ops for the same function, this seems to come down to individual preference. Perhaps there’s a middle ground we could explore. For example, Torch defines both clamp and clip operations, which perform the same function, with clip simply acting as an alias for clamp:

	•	[torch.clamp](https://pytorch.org/docs/stable/generated/torch.clamp.html)
	•	[torch.clip](https://pytorch.org/docs/stable/generated/torch.clip.html)

This approach could offer flexibility without adding redundancy.

Reply:
@sdjordjevicTT Thank you, I will share this with the team then so we are all in sync. If you want I can make a page in the docs similar to [this one](https://docs.tenstorrent.com/tt-mlir/adding-an-op.html) or maybe even under that one, and then you guys can add to it as you continue working on design improvements for the passes.

Regarding having one or multiple TTIR ops, we've discussed offline and agreed to go with a single op in TTIR.

Reply:
Resolving this conversation now to unblock @uazizTT, since this PR is in compliance with the guidelines above.